### PR TITLE
fix(alarm): stop meant snooze inside the ringing minute, and four more

### PR DESCRIPTION
**Pull request by rryles**
Five defects in the Alarm service, found by code-reading and each confirmed against the sources. The headline one is user-visible: **Stop, pressed inside the alarm's own minute, rang the alarm straight back and left a snooze pending** — so Stop behaved as Snooze for most of the ringing minute.

The ringing screen's R1/R2 mapping (top-right = Snooze, bottom-right = Stop, as swapped by 7a556a3e for apps-v1.3.0) is correct and consistent across `RingingView`, `RingingViewBase` and `ButtonsBase`. It is untouched here.

## 1. Stop inside the firing minute re-armed the alarm

`mSnoozedAlarms` was doing two unrelated jobs: holding pending snoozes, *and* acting as the latch that kept `checkAlarms()` from ringing the same minute twice (`!isSnoozed(alarm)`). Stop clears that list, and `Service::run()` calls `execute()` again on the very next loop iteration — so a repeating alarm re-matched its own minute, rang again, and armed a fresh snooze at +5 min.

The window is nearly the whole minute (the alarm fires within a second or two of the boundary, since `execute()` sleeps `(60 - tm_sec) * 1000`), so most users were inside it. Only repeating alarms were affected: a `REPEAT_NO` alarm has `on` cleared by its first trigger and cannot re-match.

Fixed by splitting the two jobs. `mFiredAlarms` latches "this alarm already rang in this clock minute", is keyed on the epoch minute, and is deliberately **not** cleared by Stop — while the snooze list now means only "a ring is owed later".

Worth noting for the field report: because the re-ring reaches the GUI as an `ActivatedAlarm`, the practical symptom was not only "it comes back in five minutes" but "the ringing screen reappears immediately, and pressing Stop again does it again until the minute rolls over". Snooze, by contrast, left the entry in place and went quiet — which reads as *Stop*. That is a much closer fit to the report than I expected going in, but the apps-v1.3.0 mapping change is still the more likely half of it, and this PR does not claim to close it.

## 2. Snooze re-trigger used exact hh:mm equality

`nextTriggerHour == currentHour && nextTriggerMinute == currentMinute`: miss that one minute and the entry never fired *and* was never erased — and a stuck entry kept `isSnoozed()` true, suppressing its own alarm on every following day until the app was unloaded.

Deadlines are now absolute timestamps compared with `>=`, measured on UTC so a timezone change (written twice on every BLE connect) cannot shift a pending snooze. They are anchored to the start of the minute, because the service only wakes on minute boundaries: a raw `now + 300 s` set at `:02` is missed by the `:00` wake and every five-minute snooze silently becomes six.

Both out-of-range cases are handled rather than left to stick:

- **missed by more than the snooze interval** (the watch was off, or the clock stepped forward) — dropped, since ringing then would ring at the wrong time;
- **clock stepped backwards** past any plausible deadline — re-anchored, so no entry can outlive its usefulness.

## 3. A one-off alarm's pending snooze was orphaned the moment it fired

For `REPEAT_NO`, the trigger clears `alarm.on`, and `removeObsoleteSnoozedAlarms()` drops any snooze whose alarm is missing *or* `!on`. The pending snooze was therefore already condemned: any later `saveAlarmList()` or `disableAlarm()` silently cancelled it. In practice — snooze a one-off alarm, then toggle or edit anything in the app within those five minutes, and the snooze never returns.

Such a snooze is now armed **detached** and survives its parent being switched off *by the trigger itself*. The user's own decision to switch an alarm off, or to delete or re-time it, still cancels the snooze.

While fixing this I found a related trap worth a comment in the code: the GUI reads `Alarm::on` as "this alarm is ringing" (`Model::switchToNextPriorityScreen`), so a one-off's re-ring must carry `on` set even though the alarm is now off. The old code got this right only by accident of statement order; `armSnooze()` now sets it explicitly, and a test guards it.

## 4. Pending snoozes were RAM-only

Per `Docs/service-lifecycle.md` §7, reboot, OTA and an update-reboot deliver **no** `COMMAND_APP_STOP` at all, and shutdown / factory reset / MSD / low battery tear the app down. There is no "last chance to save", so persist-as-you-go is the only correct pattern.

Snoozes are now written whenever they change and validated against the alarm list on load. Two deliberate choices:

- **Their own file** (`snoozes.json`). Snooze data can then never crowd the alarm list out of the shared 2 KB serialisation buffer, a write failure on one cannot cost the user the other, and `alarms.json` keeps its format.
- **`uint32` epoch seconds** for the deadline. `JsonStreamWriter`'s `int64_t`/`uint64_t` overloads cast to `double` and print with `%g` (there is a `TODO` on both), which rounds a timestamp to six significant digits — about a quarter-hour of slop. Caught by a test that failed with `"next_at":1.76725e+09`. The 32-bit path is exact and holds epoch seconds until 2106. **Flagging it as a trap for anything else that persists a 64-bit value**, but not fixing it here: it would change serialisation behaviour for existing consumers and deserves its own PR.

## 5. The service asked to be woken every minute for ever

`mGuiStarted` clears only on `COMMAND_APP_NOTIF_GUI_STOP`, and navigating away *suspends* a GUI without sending one (§6), so the "no active alarms and GUI not started" exit could not fire again after the first GUI run.

I have **not** made the service exit under a suspended GUI, and I think the original branch is closer to right than it looks: `mGuiStarted` accurately tracks a GUI being **loaded**, and §5.3 is explicit that a service returning from `run()` takes its GUI down **without notice** — no `onStop()`, nothing. Exiting there would close the app under a user who had merely glanced at a notification. §6 puts that decision with the GUI ("the GUI has to take that decision and exit; no one else will"), and no example app exits on suspend today.

So what this PR fixes is the part that is unambiguously wrong — the **cost** of the wait. `execute()` now returns `kNoWork` when no alarm is enabled and no snooze is pending, so the service blocks on `getMessage()` instead of waking every 60 s for the life of the app; the exit test's startup grace still bounds the wait before any GUI appears. The residency itself is now documented at the exit test rather than left looking like an oversight.

**The remaining half is a product decision I did not want to take unilaterally**: should navigating away from Alarm close it? Two ways to do it if the answer is yes — the GUI exits in `onSuspend()`, or it tells the service about focus with a custom message and the service exits after a grace period (30 s would match the GUI's existing on-screen idle timeout). Happy to follow up with either. Note the app is `APP_AUTOSTART On`, so with an alarm enabled the service is resident from boot regardless; this only bites in the session where the user opened the app with no alarms set.

## The constraint on all of it

`execute()` is called far more often than once a minute — every message wakes `getMessage()` and restarts the loop, and a resident service is handed events it never asked for. `checkAlarms()` is therefore idempotent within a clock minute throughout, and a test rings 60 consecutive seconds' worth of calls through it to prove one ring.

## Also checked

**Timer does not repeat this pattern.** Its `execute()` acts only on `RUNNING` and the `RUNNING -> FIRED` transition is one-way, so `stop()` (which sets `IDLE`) cannot be undone by the next `execute()`. Its exit test shares the `mGuiStarted` shape described in §5, but nothing re-arms behind it.

**The single-alarm Stop path had the same bug** as #1 and is fixed with it, though `ALARM_STOP` / `ALARM_SNOOZE` are currently unreachable — `Model::stopAlarm()` and `snoozeAlarm()` only ever send the `*_ALL` variants.

I did not chase the `transitionBegins()` / `gotoXScreenNoTransition()` re-entrancy; `evaluatePendingScreenTransition()` clears the pending callback after `execute()` returns, so it is harmless.

## Verification

- **28 new host tests** in `Tests/Host/apps/Alarm/AlarmManager_test.cpp`, one per defect plus the edge cases (missed minute, late-by-hours, clock step backwards, restart round-trip, corrupt snooze file, one-off detachment, the `on`-as-ringing contract, wake-up cadence). Full suite: **365 passed, 0 failed**.
- The ARM `.uapp` compiles, links and packs (`Alarm_1.0.0-dirty.uapp`, 216 KB).
- The TouchGFX simulator builds (`Application.exe`), which is the other consumer of both changed sources.

Not reproduced on hardware. The tests drive the exact sequences the defects need, including the Stop-inside-the-minute one, but a bench confirmation of #1 before and after would still be worth having.

## Summary by CodeRabbit

* **New Features**
  * Snoozed alarms now persist across application restarts.
  * Alarm scheduling is more reliable across clock changes and delayed wake-ups.
  * One-time alarms can be snoozed without becoming recurring alarms.
  * The alarm service improves startup responsiveness when no alarms are pending.

* **Bug Fixes**
  * Prevented duplicate alarm firing within the same minute.
  * Improved handling of corrupted, incomplete, or temporarily unavailable alarm data.
  * Improved recovery and reactivation of pending snoozes.

* **Tests**
  * Added comprehensive coverage for alarm firing, repeats, stopping, snoozing, persistence, and wake-up scheduling.

### CODE REVIEW

File: Examples/Apps/Alarm/Software/Libs/Sources/Service.cpp [hunk 1]
```diff
@@ -46,7 +63,15 @@ void Service::run()
...
-        uint32_t sleepTime = mAlarmManager.execute(getLocalTime());
+        const Instant now = getNow();
+        uint32_t sleepTime = mAlarmManager.execute(now.local, now.utc);
+
+        // execute() answers kNoWork when nothing is armed, which would park us
+        // on getMessage() for ever. That is what we want once the app is
+        // established, but not before the exit test below has had its say.
+        if (!mGuiStarted && sleepTime > kStartupGraceMs) {
+            sleepTime = kStartupGraceMs;
+        }
```
[the remark this answers is not in the document]
Comment (rryles) [reply Examples/Apps/Alarm/Software/Libs/Sources/Service.cpp]: Good catch — this is right, and it was a real regression in the change. Fixed in 48fa08e2.

Confirmed the mechanism before fixing: the Alarm app is `APP_AUTOSTART On`, so "alarm armed, no GUI loaded" is not an edge case, it is the service's ordinary resident state from boot. In that state `execute()` returns the wait until the next minute (up to 60000 ms) and the old condition cut it to 5000 ms, so the service woke ~12 times a minute for as long as any alarm stayed set. And as you say, the exit test cannot fire there because `hasActiveAlarms()` is true — so not one of those extra wake-ups could do any work. A 12× wake cadence in the common case, which is precisely the cost the `kNoWork` change was made to remove.

Applied your condition rather than a variation on it: `kNoWork` is the only value that would park `getMessage()` for ever, and it means exactly what the exit test acts on, so it is the only one worth capping. Every other value is a real deadline and is now left alone. Comment updated to say why, so the next reader doesn't "tidy" it back.

Re-verified: host suite 365/365, and `AlarmService.elf` recompiles and links for ARM. Note the cap itself has no regression test — the wake-up cadence is covered at the manager level (`AsksForNoWakeUpWhenNothingIsArmed`, `AsksForTheNextMinuteWhileAnAlarmIsArmed`), but `Service::run()`'s loop has no host harness to drive it.

**Comment by rryles** [comment]
## Review pass over this PR: one real defect fixed, six edge cases logged

A review pass over the branch turned up one genuine bug in my own change, fixed in a708f757, plus six smaller edge cases I have deliberately **not** fixed here. Logging them so they are on the record rather than lost.

### Fixed: a spent snooze could swallow its own alarm for a whole day

The alarm scan skips any alarm with a snooze pending, and it ran *before* the snooze sweep — so an entry the sweep was about to drop as unringable was still enough to skip the alarm. An alarm's minute does not come round again until the next day, so that cost a full day's alarm.

Reachable the ordinary way: a daily alarm rings, the user snoozes, the watch is off over the deadline and comes back inside the alarm's own minute the next day. `load()` restores the snooze, the first `execute()` skips the alarm because a snooze is "pending", the sweep then drops that snooze as hours late, and the next wake is a minute too late. **No ring that day.** Persisting snoozes is what made this reachable, so it arrived with the first commit of this PR.

Settling is now its own pass (`settleSnoozes()`) running before the alarm scan, so anything that can no longer ring on time is dropped or re-anchored while it still cannot matter. The re-trigger loop is left with one question — is this entry due. Three tests added, including a due snooze sharing a minute with an unrelated alarm, which had no coverage before. 368 host tests pass; `AlarmService.elf` rebuilds for ARM.

### Not fixed here — noted for a follow-up

Each of these is narrower than the defects this PR set out to fix, and none is a regression against `main`. Happy to take any of them in this PR if a reviewer would rather see them dealt with now.

1. **`detached` is sticky.** A one-shot rings and switches itself off; if the user turns it back on (for tomorrow) and then off again, that second, genuinely user-initiated "off" no longer cancels the snooze. The stated contract — a user switching an alarm off cancels its snooze — does not hold for a detached entry. Cheap fix: clear `detached` in `saveAlarmList()` for any entry whose incoming alarm is `on`, re-attaching it to its parent.
2. **A transient `alarms.json` read failure at boot discards every persisted snooze.** `load()` validates restored snoozes against `mAlarms` without checking that `loadFromFile()` succeeded; on failure `mAlarms` is empty, so every snooze looks orphaned, is dropped, and is persisted as dropped. Guard the validate-and-persist step on the alarm load having succeeded.
3. **A large backward clock step re-anchors to an arbitrary wall time.** If the RTC comes up a day behind (battery pull) with a snooze restored, it re-anchors to "five minutes from now" — a ring five minutes after boot, whatever the time. Worth bounding: re-anchor only for a small step, drop otherwise.
4. **The `kMaxSnoozes` refusal is silent.** With 8 snoozes tracked a 9th alarm rings with no snooze armed, so Snooze does nothing; and for a one-shot with no GUI loaded, `hasActiveAlarms()` is false straight after the ring, so the service can exit before `COMMAND_APP_NOTIF_GUI_RUN` arrives. Needs 9 alarms firing inside 20 minutes.
5. **Defect 1 re-emerges across a reboot inside the ringing minute.** The fired latch is RAM-only by design and Stop persists an empty snooze list, so Stop at 07:00:10 followed by a reboot completing at 07:00:50 rings again. Persisting the latch would cost a flash write on every ring; I do not think it is worth it, but it is a real hole in the fix.
6. **`persistSnoozesIfDirty()` clears the dirty flag on a failed write.** If `open()` fails (rather than the write truncating), the previous file content survives — so Stop followed by a reboot within five minutes can ring a stopped alarm. Where the write truncates first (FatFs `CREATE_ALWAYS`, and the test fixture) the outcome is a lost snooze instead, which is benign.

**Comment by rryles** [comment]
Follow-up: minor findings 1 and 2 from the list above are now fixed in f57f4a1d. Four remain open and are unchanged.

**1. Sticky `detached`** — a detached snooze stayed detached for ever, so if the user switched a fired one-shot back on and then off again, that second genuinely user-initiated "off" was ignored and the snooze still rang. A list save whose incoming alarm is enabled now re-attaches the snooze to it, and it is cancelled like any other alarm's snooze thereafter. Re-enabling on its own still keeps the snooze — that is the case the flag exists for, and there is now a test for each direction. Confirmed the new test fails before the change (rings twice).

**2. Snoozes destroyed by an unreadable alarm list** — `load()` validated restored snoozes against a list it never checked had loaded, and `mAlarms` is empty whether `alarms.json` is absent or unreadable, so a transient read failure made every pending snooze look orphaned *and persisted the loss*. Validation now runs only against a list that actually loaded; otherwise the snoozes are kept unvalidated, since each carries its own time and effect and anything no longer wanted is dropped by the next list save or by the stale-snooze grace. Losing user state on the strength of a file we failed to read is the worse trade.

Still open, deliberately: **3** (large backward clock step re-anchors to an arbitrary wall time), **4** (silent `kMaxSnoozes` refusal), **5** (defect 1 re-emerges across a reboot inside the ringing minute — the fired latch is RAM-only by design, and persisting it would cost a flash write per ring), **6** (`persistSnoozesIfDirty()` clears the dirty flag on a failed write).

371 host tests pass; `AlarmService.elf` rebuilds for ARM.

**Comment by rryles** [comment]
@coderabbitai review

File: Examples/Apps/Alarm/Software/Libs/Sources/AlarmManager.cpp [hunk 2]
```diff
@@ -23,25 +24,62 @@ AlarmManager::~AlarmManager()
...
+    const bool alarmsLoaded = loadFromFile(mAlarms);
 
     LOG_DEBUG("Alarms loaded\n");
     dump(mAlarms);
 
+    // Snoozes outlive a restart, but only as far as their alarm does: one whose
+    // alarm has been deleted, or switched off by the user, is not owed a ring.
+    loadSnoozesFromFile();
+
+    if (alarmsLoaded) {
```
[the remark this answers is not in the document]
Comment (rryles) [reply Examples/Apps/Alarm/Software/Libs/Sources/AlarmManager.cpp:36]: Right, and thank you — this is the same defect I fixed one commit earlier, for a partial failure instead of a total one, and I missed the partial case. Fixed in 3c2b1b16.

Verified the mechanism before fixing: `parseJSON()` has seven `continue` paths that skip a single malformed record, and it returns `true` regardless — so valid JSON with one bad field yields a partial `mAlarms` reported as a clean load. `load()` then trusted it, `removeObsoleteSnoozedAlarms()` convicted the missing alarm's snooze as orphaned, and `persistSnoozesIfDirty()` made the loss permanent.

`parseJSON()` now returns false when it parsed fewer records than the array declared. The records that *did* parse are still returned and still shown — the list is simply no longer treated as complete enough to convict a pending snooze on, which routes it into the same "keep them unvalidated" branch as an unreadable file. I preferred that over a separate partial-result signal because every caller of the completeness answer wants exactly this question, and there is only one caller.

Added the test you asked for, near enough exactly: valid JSON with `time_h: 99` on one of two alarms, asserting the skipped record's snooze survives, stays on disk, and still rings. Confirmed it fails without the fix — `snoozes.json` emptied and no ring — rather than assuming it would. 372 host tests pass and `AlarmService.elf` rebuilds for ARM.

One thing your finding points at that I have deliberately left alone, so it is on the record rather than quietly skipped: a partial parse also means the *alarm list itself* is short, and the next `saveAlarmList()` from the GUI will persist that truncated list, losing the unreadable alarm for good. That predates this PR and is a wider question than the snooze work (the app has no "some of your alarms could not be read" state to show), so it wants its own change rather than being folded in here.
